source/common/router/config_impl.cc

```diff
@@ -1117,6 +1109,17 @@ std::string RouteEntryImplBase::newPath(const Http::RequestHeaderMap& headers) c
     final_port = redirect_config_->port_redirect_.c_str();
   } else {
     final_port = "";
+    const absl::string_view current_path = headers.getHostValue();
+    const auto port_start = Http::HeaderUtility::getPortStart(current_path);
+    if (port_start != absl::string_view::npos) {
+      const auto deduced_port = current_path.substr(port_start);
+      // use deduced port only when scheme did not change to avoid using port 80 when scheme
+      // changed from http to https and to avoid using port 443 when scheme changed from
+      // https to http.
+      if (!newSchemePortMismatch(headers.getForwardedProtoValue(), deduced_port, final_scheme)) {
```

Comment:
Is the negative case covered in tests below?

Reply:
It is covered because without this check previously existing tests failed.